// obigo-js-ui-rnbs/utils/event.js
function getEvent (e) {
  return !e ? window.event : e
}

export function position (e) {
  let posx, posy
  e = getEvent(e)

  if (e.touches && e.touches[0]) {
    e = e.touches[0]
  }
  else if (e.changedTouches && e.changedTouches[0]) {
    e = e.changedTouches[0]
  }

  if (e.clientX || e.clientY) {
    posx = e.clientX
    posy = e.clientY
  }
  else if (e.pageX || e.pageY) {
    posx = e.pageX - document.body.scrollLeft - document.documentElement.scrollLeft
    posy = e.pageY - document.body.scrollTop - document.documentElement.scrollTop
  }

  return {
    top: posy,
    left: posx
  }
}

export function targetElement (e) {
  let target
  e = getEvent(e)

  if (e.target) {
    target = e.target
  }
  else if (e.srcElement) {
    target = e.srcElement
  }
  return target
}

export function getMouseWheelDirection (e) {
  e = getEvent(e)
  return Math.max(-1, Math.min(1, (e.wheelDelta || -e.detail)))
}

export function triggerBodyClickEvent () {
  const e = new MouseEvent('click',{
    'view': window,
    'bubbles': false,
    'cancelable': true
  })
  window.dispatchEvent(e)
}

export function mouseEventPolyfill () {
  try {
    new MouseEvent('test')
    return false; // No need to polyfill
  } catch (e) {
    // Need to polyfill - fall through
  }

  // Polyfills DOM4 MouseEvent
  var MouseEvent = function (eventType, params) {
    params = params || { bubbles: false, cancelable: false }
    var mouseEvent = document.createEvent('MouseEvent')
    mouseEvent.initMouseEvent(eventType, params.bubbles, params.cancelable, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null)

    return mouseEvent
  }

  MouseEvent.prototype = Event.prototype

  window.MouseEvent = MouseEvent
}
