// obigo-js-ui-rnbs/features/deviceModel.js
import application from './appManager'
const MODEL_NAME = {
  'RENAULT7P': 'renault7p',
  'RENAULT9P': 'renault9p',
  'DACIA7P': 'dacia7p',
  'NISSAN9P': 'nissan9p'
}
const WHEEL_POSITION = {
  'RHD': 'rhd',
  'LHD': 'lhd'
}
const util = (application) ? window.applicationFramework.util : undefined
const model = (util && util.getDeviceModel) ? util.getDeviceModel() : (window.$model) ? window.$model : MODEL_NAME.RENAULT7P // nissan9p renault7p renault9p dacia7p
const wheelPosition = (util && util.getSteeringWheelPosition) ? (util.getSteeringWheelPosition() === 1) ? WHEEL_POSITION.RHD : WHEEL_POSITION.LHD : WHEEL_POSITION.LHD
const defaultOptions = {
  list: {
    listIndicator: (model === MODEL_NAME.NISSAN9P),
    hideDummyItem: (model === MODEL_NAME.NISSAN9P),
    scrollbars: (model !== MODEL_NAME.NISSAN9P),
    fadeScrollbars: (model !== MODEL_NAME.NISSAN9P),
    bounce: (model !== MODEL_NAME.NISSAN9P),
    preventBounce: (model === MODEL_NAME.NISSAN9P),
    enableJogUpDown: (model === MODEL_NAME.NISSAN9P)
  }
}
export function install (_Vue, opt) {
  _Vue.prototype.$model = {
    name: model,
    wheelPosition: wheelPosition,
    options: {
      list: (opt && opt.list) ? opt.list : defaultOptions.list
    }
  }
}
